### PR TITLE
Add i18n plugin

### PR DESCRIPTION
**Pull request by micgro42**
In the compositionApi we cannot access `this`, so using a global instance method does not work. The established way to deal with this in Vue 3 is to use provide/inject.

Using provide/inject works well, with the one exception that requires a fallback or will be typed `Messages|undefined`. This leaves a couple of non-ideal solutions:
1. assert it as non-null it with `!`
2. cast it with `as Messages`
3. have some if-guard that throws an Error
4. Provide a simple fallback

I went with option 4 and extended the original Messages class from [Data Bridge] with a fallback if no repository is provided. This also makes the initialization in the index.html easier and still allows for a more elaborate mock in the future, if we want it.

[Data Bridge]: https://github.com/wikimedia/Wikibase/blob/f73e98507da6e494ffaa1fe9c15b787a527ba33a/client/data-bridge/src/presentation/plugins/MessagesPlugin/Messages.ts

Bug: T301570

### CODE REVIEW

File: src/components/HelloWorld.vue
Comment (itamargiv) [inline src/components/HelloWorld.vue:9]: I like that we have adopted the composition API in full. This is a great use case example of provide and inject.

File: src/plugins/MessagesPlugin/Messages.ts
Comment (itamargiv) [inline src/plugins/MessagesPlugin/Messages.ts:38]: Unescaped in this context means the raw text string? What is the use case example for unescaped text if we don't plan on using it as `innerHTML`?

**Review by itamargiv — COMMENTED** [review]
This looks excellent. I have one question, just to clarify the usage context of the `MessagesRepository` interface, but in any case, this is a great adapter :+1:

File: src/plugins/MessagesPlugin/Messages.ts
Comment (micgro42) [reply src/plugins/MessagesPlugin/Messages.ts:38]: We want to use this in templates directly with `{{}}`, e.g. `<label>{{ messages.getUnescaped( 'label-text ) }}</label>` where it would be double-escaped with just using `.get()`.

Comment (itamargiv) [reply src/plugins/MessagesPlugin/Messages.ts:38]: Aha, I see, thank you!

File: src/components/HelloWorld.vue
Comment (lucaswerkmeister) [inline src/components/HelloWorld.vue:17]: This discussion doesn’t have to block merging, but I wonder what the benefit of the `KEYS` enum is here? I think it would be easier to find the message used here in `i18n/en.json`, or to find all the places in the code using a certain message, if we removed this level of indirection and just had:

```html
 	<p>i18n example: {{ $messages.get( 'wikibaselexeme-newlexeme-lemma' ) }}</p>
```

We could still have tsc check that only known message keys are used by declaring the message key as `keyof MessageKeys` or something like that.

**Review by lucaswerkmeister — DISMISSED** [review]

File: src/main.ts
Comment (itamargiv) [reply src/main.ts:16]: I think we need to distinguish between injecting configurations and injecting services. Config for example is probably also going to receive a URL for the form to be sent to

Comment (micgro42) [reply src/main.ts:16]: To me, I think I would treat "config" and "services" differently. But this is not intended to be the final form of that signature yet, I would probably collect the `messageRepo` and `mw.api` into one object

**Comment by lucaswerkmeister** [comment]
Since #27 was just merged, it would be nice to have a unit test for the `provide()` too – is that possible in VTU?

File: src/components/HelloWorld.vue
Comment (micgro42) [reply src/components/HelloWorld.vue]: Yes, I like that more, and we even still have IDE autocompletion! 🙌

File: src/plugins/MessagesPlugin/Messages.ts
Comment (lucaswerkmeister) [inline src/plugins/MessagesPlugin/Messages.ts:11]:
```suggestion
 * `this.$messages.get( 'wikibaselexeme-newlexeme-lemma' )`
 * `this.$messages.getText( 'wikibaselexeme-newlexeme-lemma' )`
```

**Comment by micgro42** [comment]
> Since #27 was just merged, it would be nice to have a unit test for the `provide()` too – is that possible in VTU?

I have not yet found a way to do that. Though, I'm having trouble inspecting the object returned by `createAndMount()`. It seems to be a Proxy of a Proxy? o.O

File: src/plugins/MessagesPlugin/MessageKeys.ts
Comment (itamargiv) [inline src/plugins/MessagesPlugin/MessageKeys.ts:2]: Wait, isn't `MessageKeys` used to annotate types for the plugin's `get()` and `getUnescaped`? I'm not sure a literal alias is the way to go here, unless I'm missing something

Comment (micgro42) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: Mh, I maybe should have added more. My goal is to extend this as a list of all allowed messages:

```ts
type MessageKeys =
	'wikibaselexeme-newlexeme-lemma'
	| 'foo'
	| 'bar';

```

Comment (micgro42) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: That is, the type is intended to be "one of these exact literal strings"

Comment (micgro42) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: > Ah, yeah I got that, but I thought this would be derived from an object similar to what we added in PR #33 (exported from a file ofc, and not just ad-hoc in index.html):
> 
> https://github.com/wmde/new-lexeme-special-page/blob/5492aa7ddae6ec22b56d8d499f650d604fb5d5e2/index.html#L13-L15

But, these messages are being stored outside of this git repository, in https://github.com/wikimedia/mediawiki-extensions-WikibaseLexeme/blob/master/i18n/en.json

I don't think duplicating this copy is a good idea. They are bound to diverge.

Comment (itamargiv) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: Yes, I also understand that. I just think having a version of the copy available for local development could be useful, especially for visual tasks like copy on a button. It doesn't have to stay in sync.

Comment (micgro42) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: Yeah, having mocks for the copy makes sense, I agree with that. But mocks are no source for types for production code. Rather the other way around, the mock could be typed with the production types:

```ts
 const mockMessages: Record<MessageKeys, string> = { 
 	'wikibaselexeme-new-lexeme-submit': 'Create Lexeme', 
 };
 ```
 
 or something like that.

File: src/main.ts
Comment (lucaswerkmeister) [reply src/main.ts:16]: Alright, I guess we can do that later.

File: src/plugins/MessagesPlugin/MessageKeys.ts
Comment (itamargiv) [reply src/plugins/MessagesPlugin/MessageKeys.ts:2]: > But mocks are no source for types for production

Agreed, just wanted to prevent having to synchronize the same keys everywhere.

**Review by lucaswerkmeister — APPROVED** [review]
LGTM, should be enough to unblock #33 even if we change the details later.

**Review by itamargiv — APPROVED** [review]

**Pull request merged**
